apis/cluster/cluster.go

```diff
@@ -29,28 +32,197 @@ const (
 	CloudPacket       = "packet"
 )
 
+// Cluster is what we use internally in Kubicorn.
+// This used to be the internal Kubicorn API but
+// we are now migrating to the official cluster API (below)
+// This represents a Kubernetes Cluster
 type Cluster struct {
 	metav1.TypeMeta   `json:",inline"`
 	metav1.ObjectMeta `json:"metadata,omitempty"`
-	Name              string         `json:"name,omitempty"`
-	Project           *Project       `json:"project,omitempty"`
-	CloudId           string         `json:"cloudId,omitempty"`
-	ServerPools       []*ServerPool  `json:"serverPools,omitempty"`
-	Cloud             string         `json:"cloud,omitempty"`
-	Location          string         `json:"location,omitempty"`
-	SSH               *SSH           `json:"SSH,omitempty"`
-	Network           *Network       `json:"network,omitempty"`
-	Values            *Values        `json:"values,omitempty"`
-	KubernetesAPI     *KubernetesAPI `json:"kubernetesAPI,omitempty"`
-	GroupIdentifier   string         `json:"groupIdentifier,omitempty"`
-	Components        *Components    `json:"components,omitempty"`
-
-	// Add the official Kubernetes Cluster API for us to begin transitioning (lol) to the API
+
+	// Name is the publically available name of the Cluster
+	Name string
+
+	// ClusterAPI is the official Kubernetes cluster API
+	// We have this wrapped in a larger struct to support the transition
+	// and possibly allow for more values downstream
 	ClusterAPI *clusterv1.Cluster
+
+	// ControlPlane is the control plane MachineSet
+	ControlPlane *clusterv1.MachineSet
+
+	// MachineSets are a subset of worker machines
+	MachineSets []*clusterv1.MachineSet
+}
+
+// ProviderConfig is a convenience method that will attempt
+// to return a ControlPlaneProviderConfig for a cluster.
+// This is useful for managing the legacy API in a clean way.
+// This will ignore errors from json.Unmarshal and will simply
+// return an empty config.
+func (c *Cluster) ProviderConfig() *ControlPlaneProviderConfig {
+	//providerConfig providerConfig
+	raw := c.ClusterAPI.Spec.ProviderConfig
+	providerConfig := &ControlPlaneProviderConfig{}
+	err := json.Unmarshal([]byte(raw), providerConfig)
+	if err != nil {
+		logger.Critical("Unable to unmarshal provider config: %v", err)
```

Comment:
Would it be a good idea to return error here as well, or that would just create a big overhead when using `ProviderConfig()`?
Also, is `logger.Critical` going to stop the code execution?

Reply:
No, `logger.Critical` will not stop the code execution. https://github.com/kubicorn/kubicorn/blob/master/pkg/logger/logger.go#L70-L86

Returning the error would be smart, but would introduce a lot of problems with implementing the code in the rest of the program...

A few hundred references, each needing to check for an error that shouldn't ever happen.

Furthermore this is temporary, so I think it's okay